internal/staging_lockfile/staging_lockfile_test.go

```diff
@@ -0,0 +1,170 @@
+package staging_lockfile
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/containers/storage/pkg/reexec"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestMain(m *testing.M) {
+	if reexec.Init() {
+		return
+	}
+	os.Exit(m.Run())
+}
+
+// subTryLockPath starts a child process.
+// The caller must call Wait() on the returned cmd.
+func subTryLockPath(path string) (*exec.Cmd, io.ReadCloser, error) {
+	cmd := reexec.Command("subTryLockPath", path)
+	rc, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, nil, err
+	}
+	if err := cmd.Start(); err != nil {
+		return nil, nil, err
+	}
+	return cmd, rc, nil
+}
+
+// subTryLockPathMain is a child process which tries to opens the StagingLockfile,
+// If it has acquired the lock, it will unlock and delete the file.
+// Otherwise, it will print an error message to stdout.
+func subTryLockPathMain() {
+	if len(os.Args) != 2 {
+		fmt.Printf("expected two args, got %d", len(os.Args))
+		os.Exit(1)
+	}
+	tf, err := TryLockPath(os.Args[1])
+	if err != nil {
+		fmt.Printf("error opening lock file %q: %v", os.Args[1], err)
+		os.Exit(1)
+	}
+	tf.AssertLocked()
+	if err := tf.UnlockAndDelete(); err != nil {
+		fmt.Printf("error unlocking and deleting lock file %q: %v", os.Args[1], err)
+		os.Exit(1)
+	}
+}
+
+func init() {
+	reexec.Register("subTryLockPath", subTryLockPathMain)
+}
+
+func TestCreateAndLock(t *testing.T) {
+	l, _, err := CreateAndLock(t.TempDir(), "staging-lockfile")
+	require.NoError(t, err)
+
+	l.AssertLocked()
+
+	require.NoError(t, l.UnlockAndDelete())
+
+	require.Len(t, stagingLockFiles, 0)
+
+	// l.Assert* are NOT usable for determining lock state if there are concurrent users of the lock.
+	// It’s just about acceptable for these smoke tests.
+	assert.Panics(t, l.AssertLocked)
+}
+
+func TestTryLockPath(t *testing.T) {
+	lockFilePath := filepath.Join(t.TempDir(), "test-staging-lockfile")
+	l, err := TryLockPath(lockFilePath)
+	require.NoError(t, err)
+
+	l.AssertLocked()
+	require.NoError(t, l.UnlockAndDelete())
+
+	require.Len(t, stagingLockFiles, 0)
+	_, err = os.Stat(lockFilePath)
+	require.ErrorIs(t, err, os.ErrNotExist)
+}
+
+func TestCreateAndLockAndTryLock(t *testing.T) {
+	l, path, err := CreateAndLock(t.TempDir(), "locktest")
+	require.NoError(t, err)
+	defer os.Remove(path)
+
+	l.AssertLocked()
+
+	_, err = TryLockPath(path)
+	require.Error(t, err)
+
+	require.NoError(t, l.UnlockAndDelete())
```

Comment:
Also `require.Len(t, stagingLockFiles, 0)` after this, to ensure that the failing `TryLockPath` did not allocate anything.